modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+
+      :return: The number of times a given element is found in this List.
+      :rtype: `int`
+    */
+    proc count(x: eltType): int {
+      _enter();
+
+      var result = 0;
+
+      for i in 1.._size do {
+        if x == _getRef(i) then
+          result += 1;
+      }
+
+      _leave();
+      return result;
+    }
+
+    /*
+       Sort the elements of this List in place using their default sort order.
+
+       .. warning::
+
+        Sorting the contents of this List may invalidate existing references
+        to the elements contained in this List.
+    */
+    proc sort() {
+      // TODO: Need to look in the Sort module to see the API.
+    }
+
+    /*
+      Sort the items of this List in place using a comparator.
+
+      .. warning::
+
+        Sorting the elements of this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg comparator: A comparator object used to sort this List.
+    */
+    proc sort(comparator) {
+      // TODO: Need to look in the Sort module to see the API. 
+    }
+
+    /*
+      Index this List via subscript. Returns a reference to the element at a
+      given index in this List.
+
+      :arg i: The index of the element to access.
+
+      :return: An element from this List.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc const this(i: int) ref throws {
+      _enter();
+
+      //
+      // TODO: Do we want to switch to array style indexing instead?
+      //
+      if false && !_boundsCheck(i) then {
+        _leave();
+        const msg = "Bad index into List: " + i:string;
+        halt(msg);
+      }
+
+      try _boundsCheckLeaveOnThrow(i);
+      ref result = _getRef(i);
+      _leave();
+      return result;
+    }
+
+    /*
+      Produce a serial iterator over the elements of this List.
+
+      :yields: A reference to one of the elements contained in this List.
+    */
+    iter these() ref {
+
+      //
+      // TODO: I'm not even sure of what the best way to WRITE a threadsafe
+      // iterator is, _let alone_ whether it should even be threadsafe
+      // (I mean, is there even a point when reference/iterator invalidation
+      // is still a thing?
+      //
+      for i in 1.._size do {
+        _enter();
+        ref result = _getRef(i);
+        _leave();
+        yield result;
+      }
+    }
+
+    /*
+      Write the contents of this List to a channel.
+
+      :arg ch: A channel to write to.
+
+      :throws SystemError: When a write to the channel fails.
+    */
+    proc writeThis(ch: channel) {
+      _enter();
+
+      //
+      // TODO TODO TODO TODO
+      //
+      // Presently, if I try to let these calls throw, then they will trigger
+      // an error in modules/standard/IO.chpl along the lines of "error: call
+      // to throwing function writeThis without throws, try, or try!", so
+      // I am not sure what the most idiomatic way to write this is.
```

Comment:
This is issue #7261. If you use `ch <~> x` instead of `ch.write(x)` you will save error codes instead of throwing.

Reply:
Also I don't see any reason not to support reading the list or doing binary I/O. But these can be in another PR.